payments/payment_manager_test.go

```diff
@@ -0,0 +1,130 @@
+		overPayment   = big.NewInt(2000)
+
+		startingBalance = Balance{big.NewInt(1000), big.NewInt(0)}
+		onePaymentMade  = Balance{big.NewInt(980), big.NewInt(20)}
+		twoPaymentsMade = Balance{big.NewInt(960), big.NewInt(40)}
+	)
+
+	getBalance := func(m manager) Balance {
+		bal, _ := m.Balance(channelId)
+		return bal
+	}
+
+	// Happy path: Payment manager can register channels and make payments
+	paymentMgr := NewPaymentManager(testactors.Alice.Address())
+
+	_, err := paymentMgr.Pay(channelId, payment, testactors.Alice.PrivateKey)
+	Assert(t, err != nil, "channel must be registered to make payments")
+
+	Ok(t, paymentMgr.Register(channelId, deposit))
+	Equals(t, startingBalance, getBalance(paymentMgr))
+
+	firstVoucher, err := paymentMgr.Pay(channelId, payment, testactors.Alice.PrivateKey)
+	Ok(t, err)
+	Equals(t, testVoucher(channelId, payment, testactors.Alice), firstVoucher)
+	Equals(t, onePaymentMade, getBalance(paymentMgr))
+
+	signer, err := firstVoucher.recoverSigner()
+	Ok(t, err)
+	Equals(t, testactors.Alice.Address(), signer)
+
+	// Happy path: receipt manager can receive vouchers
+	receiptMgr := NewReceiptManager()
+
+	_, err = receiptMgr.Receive(firstVoucher)
+	Assert(t, err != nil, "channel must be registered to receive vouchers")
+
+	_ = receiptMgr.Register(channelId, testactors.Alice.Address(), deposit)
+	Equals(t, startingBalance, getBalance(receiptMgr))
+
+	received, err := receiptMgr.Receive(firstVoucher)
+	Ok(t, err)
+	Equals(t, received, payment)
+
+	// Receiving a voucher is idempotent
+	received, err = receiptMgr.Receive(firstVoucher)
+	Ok(t, err)
+	Equals(t, received, payment)
+	Equals(t, onePaymentMade, getBalance(receiptMgr))
+
+	// paying twice returns a larger voucher
+	secondVoucher, err := paymentMgr.Pay(channelId, payment, testactors.Alice.PrivateKey)
+	Ok(t, err)
+	Equals(t, testVoucher(channelId, doublePayment, testactors.Alice), secondVoucher)
+	Equals(t, twoPaymentsMade, getBalance(paymentMgr))
+
+	// Receiving a new voucher increases amount received
+	received, err = receiptMgr.Receive(secondVoucher)
+	Ok(t, err)
+	Equals(t, doublePayment, received)
+	Equals(t, twoPaymentsMade, getBalance(receiptMgr))
+
+	// re-registering a channel doesn't reset its balance
+	err = paymentMgr.Register(channelId, deposit)
+	Assert(t, err != nil, "expected register to fail")
+	Equals(t, twoPaymentsMade, getBalance(paymentMgr))
+
+	err = receiptMgr.Register(channelId, testactors.Alice.Address(), deposit)
+	Assert(t, err != nil, "expected register to fail")
+	Equals(t, twoPaymentsMade, getBalance(receiptMgr))
+
+	// Receiving old vouchers is ok
+	received, err = receiptMgr.Receive(firstVoucher)
+	Ok(t, err)
+	Equals(t, doublePayment, received)
+	Equals(t, twoPaymentsMade, getBalance(receiptMgr))
+
+	// Only the signer can sign vouchers
+	_, err = paymentMgr.Pay(channelId, triplePayment, testactors.Bob.PrivateKey)
+	Assert(t, err != nil, "only Alice can sign vouchers")
+
+	// Receiving a voucher for an unknown channel fails
```

Comment:
It seems like this will lead to race conditions. Quite likely that I receive a voucher before I register the channel?

Reply:
That's a good point. How would you suggest handling that case?

One idea is to optimistically store vouchers even for unregistered channels. I prefer to avoid that, since that gives someone else the permission to fill up my store with junk vouchers.

Reply:
I think we can avoid race conditions by ensuring that we register the channel before we send the prefund signature. If clients _won't send a voucher before they have received the prefund signature,_ we shouldn't get a race condition. If clients depart from this behaviour, I think it is OK for their voucher to be dropped. Someone else can still spam me with many prefund setup states / channel proposals -- but that is an existing problem which we likely need to solve independently.